Octave modules

ravenjoad opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This PR would fulfill the request in Issue #65398.

Much of the code mirrors that of lua and python, and not insignificant portions were built out of that.

This is a **very** messy branch, I know. I attempted a naive approach the first time, which _really_ did not work in my favor.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

No tests were present in [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests) for Octave, so there is nothing to do on that front.

ravenjoad committed:
octave.pkgs.symbolic: init at 2.9.0

ravenjoad committed:
octave.pkgs.control: init at 3.2.0

ravenjoad committed:
octave.pkgs.linear-algebra: init at 2.2.3

ravenjoad committed:
octave.pkgs.signal: init at 1.4.1

As of this commit, this matches the output that OpenSuSe's zypper
produces when installing the octave-forge-signal package.

ravenjoad committed:
octave.pkgs.io: init at 2.6.3

ravenjoad committed:
octave.pkgs.general: init at version 2.1.1

ravenjoad committed:
octave.pkgs.ltfat: init at 2.3.1

ravenjoad committed:
octave.pkgs.level-set: init at 2019-04-13

1) Version 0.3.0 has numerous bugs due to not being updated for nearly 5
years. So, I had to use fetchgit instead, grabbing HEAD of master.

2) level-set uses a non-standard layout for its files (no MAKEFILE in
the root of the package for octave to use), but instead has a build.sh
script to run. This script *JUST* generates a tarball that can THEN be
used for `pkg build`.
`patchPhase` performs some substitutions on this to make it generate
the tarball in a way we expect it to. This tarball ends up in the
/build directory we have available to us.

2.5) This script NEEDS automake, autoconf, and autoconf-archive to be
available. This is unnecessary for other packages because Octave has a
well-defined Makefile scheme that developers are supposed to follow
that allows Octave to handle the building of packages.
But this package breaks the mold.

3) With the tarball we use to build being available, we need to then
`cd` back to the previous location (as the script takes us from
`/build` to `/tmp`. So we go back using `cd -`.

4) Lastly, we can now use the standard `buildPhase` defined for
`buildOctavePackage` and complete the building of level-set.

Hopefully, this will be fixed in a later release, so it is easier to
maintain.

10/260 tests FAIL, where most of those 10 are due to improper usage of
the parallel package. Overall, I believe this is a reasonable amount
of passing tests to allow this to be marked as working.

SuperSandro2000 commented:
@KarlJoad please fix the merge conflict.

SuperSandro2000 commented on pkgs/development/octave-modules/general/default.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  nativeBuildInputs = [
+    pkgconfig
```
```suggestion
    pkg-config
```

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/f64fc7e94e26b119a373955d2ea1799cbc2acea9

SuperSandro2000 commented on pkgs/development/octave-modules/general/default.nix:
```diff
@@ -0,0 +1,28 @@
+{ buildOctaveLibrary
+, stdenv
+, fetchurl
+, pkgconfig
```
```suggestion
, pkg-config
```

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/f64fc7e94e26b119a373955d2ea1799cbc2acea9

SuperSandro2000 commented on pkgs/development/interpreters/octave/mk-octave-derivation.nix:
```diff
@@ -0,0 +1,69 @@
+    ]
+    ++ nativeBuildInputs;
+
+    buildInputs = buildInputs;
```
```suggestion
    inherit buildInputs;
```

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/c5fbd616593527c97c34f5c7e2fd556cee95dd25

SuperSandro2000 commented on pkgs/development/interpreters/octave/mk-octave-derivation.nix:
```diff
@@ -0,0 +1,69 @@
+  self = stdenv.mkDerivation {
+    packageName = "${fullLibName}";
+    name = "${namePrefix}-${fullLibName}";
+    src = src;
```
```suggestion
    inherit src;
```

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/c5fbd616593527c97c34f5c7e2fd556cee95dd25

SuperSandro2000 commented on pkgs/development/octave-modules/ltfat/default.nix:
```diff
@@ -0,0 +1,54 @@
+    jre
+  ];
+
+  meta = {
+    name = "The Large Time-Frequency Analysis Toolbox";
+    homepage = "https://octave.sourceforge.io/${pname}/index.html";
+    license = stdenv.lib.licenses.gpl3Plus;
+    maintainers = with stdenv.pkgs.maintainers; [ KarlJoad ];
```
```suggestion
  meta = with stdenv.lib; {
    name = "The Large Time-Frequency Analysis Toolbox";
    homepage = "https://octave.sourceforge.io/${pname}/index.html";
    license = licenses.gpl3Plus;
    maintainers = with maintainers; [ KarlJoad ];
```

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/53b5cfde654087b5ec8a4aaaf5705e9999385ead

doronbehar commented on pkgs/development/interpreters/octave/mk-octave-derivation.nix:
```diff
@@ -0,0 +1,69 @@
+    dontUnpack = true;
+
+    nativeBuildInputs = [
+      octave
```
I think this might warrant a comment? I wonder how python's packages handle cross compiling, where this choice might not be what we want..

KarlJoad replied:
I am not quite sure what you are referring to with this snippet here. What warrants a comment? What about python?

doronbehar replied:
> What warrants a comment?

In general, `nativeBuildInputs` is different then `buildInputs`. I don't know how to best explain this, but this makes a difference when we consider cross compiling. Currently, you don't differentiate between the octave derivation that should be used on the "host" platform and the "build" platform. This is not that significant since our octave is anyway only for x86_64 I think, but we might want to compose things in a way that if nixpkgs will gain support for compiling octave on other platforms, the `octave.pkgs` will be able to be cross compiled. IIRC octave has support for cross compiling upstream.

I'm not suggesting with this comment how things can be better, I'm just mentioning this here, because someone more experienced than me with cross compiling, will have a good say about this. So let's keep this thread open :).

KarlJoad replied:
AH! That's what you meant! I wasn't quite sure what you were referring to before.

When writing the mk-octave-derivation.nix file, I was mirroring python's [mk-python-derivation.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix). So I wasn't quite sure what I was doing (and I still don't).

But, octave can already be built for other architectures. It looks that Debian has an aarch64 version of octave (see [here](https://packages.debian.org/sid/arm64/octave/filelist)).

KarlJoad replied:
So instead of having `nativeBuildInputs = [ octave ]`, have `buildInputs = [ octave ]`, @FRidh?

FRidh replied:
No, keep it. Eventually for cross it would be `nativeBuildInputs = [ buildPackages.octave ]; `.

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,92 @@
+in rec {
+
+  control = callPackage ../development/octave-modules/control {
+    gfortran = gfortran;
+    autoreconfHook = autoreconfHook;
+    lapack = lapack;
+    blas = blas;
```
I don't like this, this should be the default - the same `gfortran` that's used for octave should be used here, without question, not choice. For that, you'd need something like this:

https://github.com/NixOS/nixpkgs/commit/2a51ddda6beec6ce04770b140700599d70e253a3

KarlJoad replied:
I'm sorry, but I'm not quite sure how that commit show-cases what you mean. 
Would you mind explaining the general idea for me? I understand why having a different `gfortran` would be bad, but I'm not sure why manually specifying that `gfortran` is needed for that package is particularly "bad".

doronbehar replied:
> I'm sorry, but I'm not quite sure how that commit show-cases what you mean.
> Would you mind explaining the general idea for me? I understand why having a different `gfortran` would be bad, but I'm not sure why manually specifying that `gfortran` is needed for that package is particularly "bad".

The idea is, that if in the future there will be a new octave package added, the person adding it won't have to take care of all of these deps safety issues - that the dependency the octave package will use exactly the same deps (if needed) that octave uses - those that are available in it's passthru. A user adding a new package should do nothing besides `callPackage <file.nix> { };` and whatever deps they wish to use there will be used thanks to the prior definition of `callPackage`.

That commit by `@FRidh` demonstrates how to solve this:

https://github.com/NixOS/nixpkgs/pull/106599#issuecomment-749709846

And essentially it sums up to using `lib.makeScope` instead of `lib.newScope`. I wasn't accurate when I gave the example of `gfortran`, as for this dep specifically we can put it in the scope of `callPackage` even with `newScope` if you do:

```nix
let
  callPackage = lib.newScope {
    # Other deps...
    inherit (octave) gfortran;
  };
in {
  pkg1 = callPackage file.nix { };
  # ...
}
```

But I think we should use `makeScope` never the less, to avoid having to specify deps in usages of `callPackage` as I (hopefully in an understandable manner) explained in https://github.com/NixOS/nixpkgs/pull/106599#issuecomment-749709846 .

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/a9597730631bb5f8b0086a273811a930cbb7b6a0

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,92 @@
+  };
+
+  io = callPackage ../development/octave-modules/io {
+    enableJava = true;
```
This should be conditioned based on whether octave is built with java.

KarlJoad replied:
This is a bit of a moot point, since [io-2.4.13   Release Date: 2019-10-15](https://sourceforge.net/p/octave/io/ci/default/tree/NEWS), when the explicit dependency on java was removed, as java support should be built into octave.

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,92 @@
+
+  io = callPackage ../development/octave-modules/io {
+    enableJava = true;
+    jdk = jdk;
```
Again, should be the same jdk as octave's.

KarlJoad replied:
Understood. As mentioned in https://github.com/NixOS/nixpkgs/pull/108562#discussion_r553513390, this has become a moot point for the `io` package. 
However, the `ltfat` package does require a JRE, which I can pass from octave. I will have to handle that later, when I figure out what is causing `ltfat` to have problems building.

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,92 @@
+  symbolic = callPackage ../development/octave-modules/symbolic {
+    # Need to use sympy 1.5.1 for https://github.com/cbm755/octsympy/issues/1023
+    # It has been addressed, but not merged yet.
+    pythonEnv = (let
+      overridenPython = let
+        packageOverrides = self: super: {
+          sympy = super.sympy.overridePythonAttrs (old: rec {
+            version = pkgs.python2Packages.sympy.version;
+            src = pkgs.python2Packages.sympy.src;
+          });
+        };
+      in python3.override {inherit packageOverrides; self = overridenPython; };
+    in overridenPython.withPackages (ps: [
+      ps.sympy
+      ps.mpmath
```
Can you compose this inside the expression? Again, ideally with `octave.python` which should be available in the passthru of it..

KarlJoad replied:
When using `nix repl`, I don't see any mention of `octave.python` or `octave.passthru.python`. Did you meant that I define it myself?

KarlJoad replied:
Sorry. Doing that rebase, and just figured out what you meant by `octave.passthru.python`. Working on it.

However, I'm not sure what you mean by "compose this inside the expression". Can you explain further?

doronbehar replied:
> However, I'm not sure what you mean by "compose this inside the expression". Can you explain further?

I meant that you can define `pythonEnv` inside the expression of this package, it's just another dependency.

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/276ff6e14e6cf7a56bd723cdd03cf596b6c9f476

doronbehar commented on pkgs/development/interpreters/octave/wrapper.nix:
```diff
@@ -0,0 +1,119 @@
+      done
+
+      if [ -L "$out/share/octave" ]; then
+         unlink "$out/share/octave"
```
Why not `rm`? :)

KarlJoad replied:
Mostly because all the other wrapper.nix-s (lua, python) use this same method. I am not sure of the benefit of this method compared to `rm`.

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/f3dcd0b42e5f220d93aee46d502e0cda7783e841

doronbehar commented on pkgs/development/interpreters/octave/wrapper.nix:
```diff
@@ -0,0 +1,119 @@
+
+# Create an octave executable that knows about additional packages
+let
+  env = let
+    paths = extraLibs ++ [ octave ];
+    octavePath = "${octave}";
+    octaveExecutable = "${placeholder "out"}/bin/octave-cli";
+  in (buildEnv {
```
You can get rid of the whole `let in` thing, and use all of these vars inside the arguments to `buildEnv`. Use `${octave}` instead of `${octavePath}`.

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/12d41c4d14ab005b5cb07c5ca62b2a50b5ce5305

ravenjoad committed:
octave.pkgs.arduino: init at 0.6.0

ravenjoad committed:
octave.pkgs.audio: init at 2.0.2

http://www.music.mcgill.ca/~gary/rtmidi/index.html#compiling contains
instructions for Mac OSX. The dependencies for Mac are: CoreMIDI,
CoreAudio, CoreFoundation

ravenjoad committed:
octave.pkgs.bim: init at 1.1.5

ravenjoad committed:
octave.pkgs.bsltl: init at 1.3.1

ravenjoad committed:
octave.pkgs.cgi: init at 0.1.2

ravenjoad committed:
octave.pkgs.communications: init at 1.2.2

ravenjoad committed:
octave.pkgs.data-smoothing: init at 1.3.0

ravenjoad committed:
octave.pkgs.database: init at 2.4.4

ravenjoad committed:
octave.pkgs.dataframe: init at 1.2.0

ravenjoad committed:
octave.pkgs.dicom: init at 0.4.0

ravenjoad committed:
octave.pkgs.divand: init at 1.1.2

ravenjoad committed:
octave.pkgs.doctest: init at 0.7.0

ravenjoad committed:
octave.pkgs.econometrics: init at 1.1.2

ravenjoad committed:
octave.pkgs.fem-fenics: init at 0.0.5

ravenjoad committed:
octave.pkgs.financial: init at 0.5.3

ravenjoad committed:
octave.pkgs.fits: init at 1.0.7

ravenjoad committed:
octave.pkgs.fpl: init at 1.3.5

ravenjoad committed:
octave.pkgs.fuzzy-logic-toolkit: init at 0.4.5

ravenjoad committed:
octave.pkgs.ga: init at 0.10.2

ravenjoad committed:
octave.pkgs.generate_html: init at 0.3.2

ravenjoad committed:
octave.pkgs.geometry: init at 4.0.0

ravenjoad committed:
octave.pkgs.gsl: init at 2.1.1

ravenjoad committed:
octave.pkgs.image: init at 2.12.0

ravenjoad committed:
octave.pkgs.image-acquisition: init at 0.2.2

ravenjoad committed:
octave.pkgs.instrument-control: init at 0.6.0

ravenjoad committed:
octave.pkgs.interval: init at 3.2.0

ravenjoad committed:
octave.pkgs.lssa: init at 0.1.4

ravenjoad committed:
octave.pkgs.mapping: init at 1.4.1

ravenjoad committed:
octave.pkgs.matgeom: init at 1.2.2

ravenjoad committed:
octave.pkgs.miscellaneous: init at 1.3.0

ravenjoad committed:
octave.pkgs.msh: init at 1.0.10

ravenjoad committed:
octave.pkgs.mvn: init at 1.1.0

ravenjoad committed:
octave.pkgs.nan: init at 3.5.2

ravenjoad committed:
octave.pkgs.ncarray: init at 1.0.4

ravenjoad committed:
octave.pkgs.netcdf: init at 1.0.14

ravenjoad committed:
octave.pkgs.nurbs: init at 1.3.13

ravenjoad committed:
octave.pkgs.ocl: init at 1.1.1

ravenjoad committed:
octave.pkgs.octclip: init at 2.0.1

ravenjoad committed:
octave.pkgs.octproj: init at 2.0.1

ravenjoad committed:
octave.pkgs.optics: init at 0.1.4

ravenjoad committed:
octave.pkgs.optim: init at 1.6.0

ravenjoad committed:
octave.pkgs.optiminterp: init at 0.3.6

ravenjoad committed:
octave.pkgs.parallel: init at 4.0.0

ravenjoad committed:
octave.pkgs.struct: init at 1.0.16

ravenjoad committed:
octave.pkgs.quaternion: init at 2.4.0

ravenjoad committed:
octave.pkgs.queueing: init at 1.2.7

ravenjoad committed:
octave.pkgs.sockets: init at 1.2.1

ravenjoad committed:
octave.pkgs.stk: init at 2.6.1

ravenjoad committed:
octave.pkgs.sparsersb: init at 1.0.8

ravenjoad committed:
octave.pkgs.splines: init at 1.3.3

ravenjoad committed:
octave.pkgs.statistics: init at 1.4.2

ravenjoad committed:
octave.pkgs.strings: init at 1.2.0

ravenjoad committed:
octave.pkgs.tisean: init at 0.2.3

ravenjoad committed:
octave.pkgs.tsa: init at 4.6.2

ravenjoad committed:
octave.pkgs.vibes: init at 0.2.0

ravenjoad committed:
octave.pkgs.video: init at 2.0.0

ravenjoad committed:
octave.pkgs.vrml: init at 1.0.13

ravenjoad committed:
octave.pkgs.windows: init at 1.5.0

ravenjoad committed:
octave.pkgs.zeromq: init at 1.5.2

KarlJoad commented:
> @KarlJoad please fix the merge conflict.

Fixed! :smile:

cole-h commented on pkgs/development/octave-modules/linear-algebra/default.nix:
```diff
@@ -0,0 +1,23 @@
+  pname = "linear-algebra";
+  version = "2.2.3";
+
+  src = builtins.fetchurl {
```
```suggestion
  src = fetchurl {
```

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108562` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages marked as broken and skipped:
</summary>
<ul>
<li>octavePackages.level-set</li>
<li>octavePackages.ltfat</li>
</ul>
</details>
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>octavePackages.control</li>
<li>octavePackages.io</li>
<li>octavePackages.linear-algebra</li>
<li>octavePackages.signal</li>
<li>octavePackages.symbolic</li>
</ul>
</details>

doronbehar commented on pkgs/development/octave-modules/symbolic/default.nix:
```diff
@@ -0,0 +1,39 @@
+    maintainers = with maintainers; [ KarlJoad ];
+    description = "Adds symbolic calculation features to GNU Octave";
+  };
+}).overrideAttrs (oldAttrs: rec {
```
Why building and then overriding? :)

KarlJoad replied:
I was under the impression that `overrideAttrs` evaluates the original, then evaluates the override without a build in between. I came to that conclusion because of the manual section about [`overrideAttrs`](https://nixos.org/manual/nixpkgs/stable/#sec-pkg-overrideAttrs).

The reason I need to do this is because symbolic needs a way to reference the python environment that is built for it, and the standard `buildOctaveLibrary` doesn't seem to have support for overriding its various phases to achieve what I wanted.

KarlJoad replied:
I needed the override because `symbolic` depends on being able to call `python3` from the environment (defined in `private/defaultpython.m`). Right now, I am using `substituteInPlace` to replace that reference. To do that:
1. I need to unpack the tarball (`dontUnpack = false`)
2. Perform the substitution (`buildPhase`)
3. Rebuild the tarball (`installPhase`)

KarlJoad replied:
Something that I thought of while I was working on the octave's arduino package was that `propagatedBuildInputs` doesn't actually get used in any meaningful way. See [here](https://github.com/NixOS/nixpkgs/pull/108562#issuecomment-756554745) for a further discussion of the solution I would *like* to use.

If I can get `wrap.sh` to work correctly and add `propagatedBuildInputs` to the `PATH`, then I can completely remove the override. Because, if I have `propagatedBuildInputs = [ pythonEnv ];`, then the python interpreter for that environment will be added to `PATH`, and that interpreter has all the correct wrapping needed to find its declared libraries. This removes the need for me to mess around with the tarball at all.

FRidh replied:
We should avoid using the "let's wrap everything that is in `propagatedBuildInputs` " hack in more places. If something needs to be wrapped, it best be explicit.

In https://github.com/NixOS/nixpkgs/pull/102613 I introduce `requiredPythonModules` to mark Python dependencies. Maybe you want the same here.

KarlJoad replied:
Ok. I will work on switching what I have already to `requiredOctaveModules`, mimicking Python's runtime dependencies. I wanted to start by getting the "let's wrap everything that is in `propagatedBuildInputs` working first. That way, I at least have a sanity check.

KarlJoad replied:
@FRidh I have introduced the `requiredOctavePackages` attribute and associated functions to ensure that octave packages have all their other octave package dependencies tracked separately from others.

I am deliberately leaving the "`propagatedBuildInputs` hack" in the code, because it is needed so that packages (like the python environment for `symbolic`) are properly referenced in wrappers for octave. I wanted to handle these separately because both of these are run-time dependencies, but octave package dependencies should actually be installed into Octave, like a package the user specifies. However, other runtime dependencies should not (and cannot, particularly now) be installed into octave.

doronbehar commented on pkgs/development/octave-modules/symbolic/default.nix:
```diff
@@ -0,0 +1,39 @@
+    '';
+  dontInstall = false;
+  installPhase = ''
+      mkdir -p $out
```
Missing hooks... In many other places as well. See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/17b08c0843c4d51c10e32d45a40e9fb19ea824d7

I am unsure if `symbolic`'s override needs to have those hooks present. 
What I would really prefer to do is completely change `buildPhase` and add an `installPhase` for `symbolic`.

KarlJoad replied:
I managed to get the override removed by using `makeWrapper` during the environment's building, so now there is no need to decompress the tarball anymore. 

Now, if there is a non-octave package run-time dependency, it should be placed in `propagatedBuildInputs` and a reference to those packages will be made in the wrapper that is generated in `wrapper.nix` and `wrap.sh` now.

If there is a run-time dependency on another octave package, it should be placed elsewhere (right now the name I've decided on is `requiredOctavePackages`), which will install the required package into octave alongside the one that was manually specified.

KarlJoad replied:
Now, `build-octave-package.nix` has `runHook`s in its `buildPhase` to ensure those are run, although they are not very helpful right now as I don't decompress the tarball, as octave prefers to work with tarballs.

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,91 @@
+    inherit (pkgs) lib stdenv;
+    inherit buildOctaveLibrary;
+    inherit fetchurl;
+    inherit gnuplot texinfo;
```
Even if not using `lib.makeScope`, this is where other deps from `octave` should be inherited.

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/b58f29416eee44575e039d463c7cb665555416e9

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,91 @@
+
+let
+
+  isOctaveFull = octave.enableQt;
```
You can even inherit these attributes in the `newScope` and then use them in the derivations without changing the names !

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/a9597730631bb5f8b0086a273811a930cbb7b6a0

doronbehar commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -93,6 +97,16 @@ let
+    inherit pkgs;
+    inherit (pkgs) lib stdenv fetchurl newScope;
+    octave = self;
+    inherit blas lapack gfortran python texinfo gnuplot;
```
You do use the same deps as octave here, but not accurately. Notice the difference between for example `blas` and `blas'`. I think it'd be best to not pass so many values here, and pass only `octave = self`, and put inherit the deps in the `callPackage` definition in `octave-packages.nix`, as explained in previous comment.

KarlJoad replied:
See https://github.com/NixOS/nixpkgs/pull/108562/commits/b58f29416eee44575e039d463c7cb665555416e9

KarlJoad commented:
I have a branch sitting locally right now where I have defined all the packages listed on Octave's [Community Packages](https://octave.sourceforge.io/packages.php) page. I have been rebasing it to keep it up to speed with the work done on this branch. Once we have reached agreement with how this looks, I want to merge those into here, before merging into `master`.

The reason I haven't added them here yet is because many of them have runtime dependencies that I need to account for, and I haven't gotten around to those yet.

KarlJoad commented:
To simplify the `postBuild` in `wrapper.nix`, I want to define a `wrap.sh` as well, which would be brought into scope with a `wrap-octave.nix` file. I was looking through both Python's and Lua's examples, and couldn't figure out how to make it work. I know there needs to be a call to `makeSetupHook`, but when I was working with that, bash functions would only accept a single argument, and completely ignore a second one. My first pass at this is [here](https://github.com/KarlJoad/nixpkgs/blob/octave/shell-script-helper/pkgs/development/interpreters/octave/wrap-octave.nix) and [the shell script](https://github.com/KarlJoad/nixpkgs/blob/octave/shell-script-helper/pkgs/development/interpreters/octave/wrap.sh)

If I can properly define a `wrap.sh` file, then that simplifies the building of the path to `octave_packages`, and will allow me to more easily add packages as run-time dependencies (using `propagatedBuildInputs`). Doing this would allow me to remove the `overrideAttrs` that is currently going on in `symbolic/default.nix`.
See [here](https://github.com/NixOS/nixpkgs/blob/ba410c2647f71089c23fddf86b5c9a47628e0b3a/pkgs/development/interpreters/python/wrap.sh#L105) to see what Python does in `wrap.sh`.

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,58 @@
+# Each attribute is an Octave library.
+# Expressions for the Octave libraries are supposed to be in `pkgs/development/octave-modules/<name>/default.nix`.
+
+{ pkgs
```
```suggestion
# Like python-packages.nix, top-level packages are not in the scope of the `callPacakge` 
# used for packages here, so when we do need packages from outside, we can inherit them
# from `pkgs`. 
{ pkgs
```

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,58 @@
+    control = callPackage ../development/octave-modules/control { };
+
+    io = callPackage ../development/octave-modules/io {
+      unzip = pkgs.unzip;
```
```suggestion
      inherit (pkgs) unzip;
```

doronbehar commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,58 @@
+    linear-algebra = callPackage ../development/octave-modules/linear-algebra { };
+
+    ltfat = callPackage ../development/octave-modules/ltfat {
+      fftw = pkgs.fftw;
+      fftwSinglePrec = pkgs.fftwSinglePrec;
+      fftwFloat = pkgs.fftwFloat;
+      fftwLongDouble = pkgs.fftwLongDouble;
+      portaudio = pkgs.portaudio;
```
Many packages here are also used in octave's build. So we need to add a passthru to octave and use here:

```suggestion
      inherit (octave) fftw fftwSinglePrec portaudio;
      inherit (pkgs) fftwFloat;
```

As explained for other deps previously.

doronbehar commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -1,4 +1,6 @@
-{ stdenv
+{ pkgs
```
In general, using `pkgs` like that inside expressions increases eval time. See: https://github.com/NixOS/nixpkgs/issues/38635#issuecomment-379678500

But if you wish to design `octave-packages.nix` in a way that will not put into scope all packages from `top-level.nix` (differently than what I did in https://github.com/NixOS/nixpkgs/pull/106599 ), than that's justified.

KarlJoad replied:
Interesting to learn. However, `pkgs` is needed for `octavePackages` to receive their potential inputs that are not already present in Octave. For example, the `io` package requires the `nettle` library for some applications. `parallel` needs `gnutls`, which Octave does not currently have. 

Is there a better way to make sure packages get the dependencies they need (that are not also used by Octave itself)?

doronbehar commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -59,6 +61,8 @@
 }:
 
 let
+  self = pkgs.octave;
```
Although we wish to use `pkgs` which is usually discouraged, here I think you can use:

```nix
let
  octavePackages = import ..../octave-packages.nix {
    inherit self;
    # OR:
    octave = self;
    # ...
  };
  self = stdenv.mkDerivation {
    # ...
    passthru = {
      # ...
      pkgs = octavePackages;
      # ...
    };
    # ...
  };
in self
```

And you'll use `pkgs` a bit less extensively. I don't know if it's that important, it's a matter of personal style I think, so you don't have to do this. FYI there's https://github.com/NixOS/rfcs/pull/83 were we discuss a bit how to compose packages sets in general in Nixpkgs.

doronbehar commented on pkgs/development/octave-modules/symbolic/default.nix:
```diff
@@ -0,0 +1,59 @@
+}).overrideAttrs (oldAttrs: rec {
+  dontUnpack = false;
+  buildPhase = ''
+      substituteInPlace inst/private/defaultpython.m --replace python3 ${pythonEnv}/bin/python
```
A ~year ago, I had a working octave with sympy, where the only two things I did were:

1. Wrapping octave with `--prefix PATH ${pythonEnv}/bin` and pythonEnv was similar to yours.
2. Adding the package with `pkg add_prefix` and `pkg load symbolic`.

That worked (until symbolic broke due to the new sympy). Maybe we should handle packages like symbolic differently? Perhaps by introducing a hook?

KarlJoad replied:
Right now, I have managed to find a way to not need the override at all. Instead of re-writing the python interpreter (that has an environment where sympy and mpmath are present), that particular python has been added to the `PATH`, so that octave can perform a normal look-up of `python3` in the environment variables.

In theory, this should work for any other package that also uses `PATH` to find the location of its other run-time dependencies.

FRidh reviewed (commented):
Nice to see you're taking this further!

KarlJoad commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -185,15 +196,25 @@ in mkDerivation rec {
+    withPackages = import ./with-packages.nix { inherit buildEnv octavePackages; };
+    pkgs = octavePackages;
+    interpreter = "${self}/bin/octave";
+    octPkgsPath = "share/octave/octave_packages";
```
Actually, it was an accident. I was copying Python and Lua's respective definition files for their interpreters' `passthru` attribute set, and left that in. But, because octave doesn't use a `site-packages` directory for holding these packages, I defined a second variable `octPkgsPath` that had the exact same path.

KarlJoad commented on pkgs/development/octave-modules/symbolic/default.nix:
```diff
@@ -0,0 +1,59 @@
+            src = python2Packages.sympy.src;
+          });
+        };
+      in python.override {inherit packageOverrides; self = overridenPython; };
```
Correct. I have a branch that I am currently working on, where there is a `wrap.sh` helper script that is used during environment creation. On that branch, I identified the issue and fixed it (quite literally last night).

This will be fixed when I merge that particular branch back into this PR branch.

KarlJoad commented on pkgs/development/interpreters/octave/mk-octave-derivation.nix:
```diff
@@ -0,0 +1,73 @@
+
+    meta = meta;
+  };
+in lib.extendDerivation true passthru self
```
Sorry, on my end the diff is 4 lines (68 - 71). Can you specify which line exactly? The `meta = meta` line?

KarlJoad commented:
The nixpkgs contribution manual states that tests should be run (if they don't take too long). An undocumented (in the manual) change to Octave was the introduction of a `pkg test` built-in function. The only requirement for this is that the package be **installed**.
An octave package is installed in the `build-env.nix` file, which becomes problematic. First, the `buildEnv` function doesn't have a `checkPhase` defined in its function definition. So, that means it will be part of the `postBuild` hook. Meaning, there must be a boolean value somewhere for an option to run tests. I have already determined the correct bash/octave code needed to run the tests for every package as well. However, I am not sure about how to inform the system which packages should be tested.

SuperSandro2000 commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,269 @@
+
+    fem-fenics = callPackage ../development/octave-modules/fem-fenics {
+      # I am still not super sure about this
+      dolfin = python.pkgs.fenics;
```
You use python2Packages a bit below so I would suggest to do the same here.

KarlJoad replied:
The comment was meant more as a reminder for me. When searching through nixpkgs, I noticed that there is no other package that provides `dolfin`. However, it *is* used by python's `fenics` package, which provides python bindings to `fenics`. When I build `python3.pkgs.fenics`, I **do** end up with a derivation and output of `dolfin` which contains the required `dolfin.h` header, but I have no way to get to them.

So right now, the problem is that the only way to access `dolfin` is through the python package, and I cannot access `dolfin` directly.

SuperSandro2000 replied:
We can copy the header files out of the package into a lib if you like but I am not at all familiar with octave.

KarlJoad replied:
That would be the only way to get this package (and one other one) to work, since they rely on the fenics package as being stand-alone.

KarlJoad replied:
"Moving" the library isn't an octave-specific thing. It's definition would be moved to `pkgs/development/libraries` (or another appropriate place), and then the python module can pull it in as a parameter (and perform any additional configuration there through an override). It's just that right now, there is no way to access the `fenics` package/library directly.

SuperSandro2000 commented on pkgs/development/octave-modules/zeromq/default.nix:
```diff
@@ -0,0 +1,28 @@
+  version = "1.5.2";
+
+  src = fetchurl {
+    url = "https://octave.sourceforge.io/download.php?package=${pname}-${version}.tar.gz";
```
Did you try using mirror://sourceforge instead?

KarlJoad replied:
I have not. Is that a preferred way of downloading these tarballs? Or is that because this link actually redirects the `curl` command to `mirror://sourceforge`?

SuperSandro2000 replied:
> I have not. Is that a preferred way of downloading these tarballs?

For anything from sourceforge it is. It redirect the downloads to known good mirrors instead of choosing a random one. You can find the list in here https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchurl/mirrors.nix#L19.

> Or is that because this link actually redirects the curl command to mirror://sourceforge?

curl does not know anything from that. Nix builds the url together with a link from the mirrors.

SuperSandro2000 commented on pkgs/development/octave-modules/zeromq/default.nix:
```diff
@@ -0,0 +1,28 @@
+    zeromq
+  ];
+
+  meta = with stdenv.lib; {
```
A recent change to contributing now says that this should be lib.

KarlJoad replied:
@SuperSandro2000, can you link to the change? I don't see any way to change `with stdenv.lib` to `with lib` without having to pass `lib` in somewhere else. This would also mean I don't need to pass `stdenv` to any of these package definitions. Is that what you are suggesting?

SuperSandro2000 replied:
https://github.com/NixOS/nixpkgs/pull/109455

>  This would also mean I don't need to pass stdenv to any of these package definitions. Is that what you are suggesting?

Yes

SuperSandro2000 commented on pkgs/development/octave-modules/vrml/default.nix:
```diff
@@ -0,0 +1,41 @@
+  ];
+
+  meta = with stdenv.lib; {
+    homepage = "https://octave.sourceforge.io/${pname}/index.html";
```
I would prefer not using pname here to easily click on the link but not that important.

KarlJoad replied:
Other than hard-coding in the `${pname}` of the package, I don't know an alternative to make this URL easier to resolve. Unless, that's what you're suggesting.

SuperSandro2000 replied:
I suggest exactly that. It makes it easier to open links from the source. For search.nixos.org it does not matter.

SuperSandro2000 commented on pkgs/development/octave-modules/tisean/default.nix:
```diff
@@ -0,0 +1,33 @@
+  };
+
+  buildInputs = [
+    gfortran
```
If this is used as a compiler shouldn't it go to nativeBuildInputs?

KarlJoad replied:
It probably should, yes. I when writing most of these files, I was unsure of whether to use `buildInputs` or `nativeBuildInputs`. I understand that the difference comes to the issue of cross-compilation (`nativeBuildInputs` are tools that are build using the architecture of the system performing the building, which may not be the same as the system running the package/program), but am still unsure about when to make a package a `buildInput` or a `nativeBuildInput`.

SuperSandro2000 commented on pkgs/development/octave-modules/parallel/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+  buildInputs = [
+    gnutls
+    pkg-config
```
That should probably also be a nativeBuildInput.

SuperSandro2000 commented on pkgs/development/octave-modules/fem-fenics/default.nix:
```diff
@@ -0,0 +1,32 @@
+  buildInputs = [
+    dolfin
+    fenics
+    pkg-config
```
Probably nativeBuildInputs.

KarlJoad replied:
Yes. I just need to have fenics as a library that I can access. Right now it's only present if `pythonPackages.fenics` is present, but I have no way to access `fenics`, so I will leave the `broken = true` label for now.

SuperSandro2000 commented on pkgs/development/interpreters/octave/build-env.nix:
```diff
@@ -0,0 +1,83 @@
+      fi
+
+      # Remove symlinks to the input tarballs, they aren't needed.
+      rm $out/*.tar.gz
```
That would break the build if a package is a zip.

KarlJoad replied:
Correct. However, the package that octave compiles with `pkg build` should always produce a tarball. I am not sure about a better way to handle the removal of the compressed compiled package.

SuperSandro2000 replied:
I did not know that. Then we can leave it like this.

SuperSandro2000 commented on pkgs/development/interpreters/octave/build-env.nix:
```diff
@@ -0,0 +1,83 @@
+      addPkgLocalList $out ${octave}
+
+      wrapOctavePrograms "${stdenv.lib.concatStringsSep " " packages}"
+     '' + postBuild;
```
```suggestion
       ${postBuild}
     '';
```

I think otherwise you need a final new line in your command. If I miss remembered ignore this.

KarlJoad replied:
This mirrors what [Python's `wrapper.nix`](https://github.com/NixOS/nixpkgs/blob/b3616bd96400ce0252c241d76fcafb64389defc6/pkgs/development/interpreters/python/wrapper.nix#L46) and [Lua's `wrapper.nix`](https://github.com/NixOS/nixpkgs/blob/b3616bd96400ce0252c241d76fcafb64389defc6/pkgs/development/interpreters/lua-5/wrapper.nix#L46) do. So, I just did what those did. 

Besides, this should just concatenate the strings together, so it shouldn't really change (other than the whitespace/newlines being a little messy in the final result that Nix will use).

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
+      qtsvg
+      qscintilla
+    ]
+    ++ stdenv.lib.optionals (ghostscript != null) [ ghostscript ]
```
I don't think the null check is required.

KarlJoad replied:
I am not the octave maintainer, so I have no knowledge about the check and list concatenation happening here. For this change, I would redirect to @doronbehar and @7c6f434c.

I am only focused on the building and installation of optional add-in octave packages right now.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
+    ++ stdenv.lib.optionals (!stdenv.isDarwin) [ libGL libGLU libX11 ]
+    ++ stdenv.lib.optionals stdenv.isDarwin [
+      libiconv
+      darwin.apple_sdk.frameworks.Accelerate
+      darwin.apple_sdk.frameworks.Cocoa
```
```suggestion
      Accelerate
      Cocoa
```

Please do inherit in the top-level with darwin.apple_sdk.frameworks.

KarlJoad replied:
Ping @doronbehar and @7c6f434c.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
+    ]
+    ;
+    nativeBuildInputs = [
+      pkgconfig
```
```suggestion
      pkg-config
```

KarlJoad replied:
Ping @doronbehar and @7c6f434c.

KarlJoad replied:
Fixed with a treewide commit.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
-  # See https://savannah.gnu.org/bugs/?50339
-  F77_INTEGER_8_FLAG = if use64BitIdx then "-fdefault-integer-8" else "";
+    # See https://savannah.gnu.org/bugs/?50339
+    F77_INTEGER_8_FLAG = if use64BitIdx then "-fdefault-integer-8" else "";
```
```suggestion
    F77_INTEGER_8_FLAG = lib.optionalString use64BitIdx "-fdefault-integer-8";
```

KarlJoad replied:
Ping @doronbehar and @7c6f434c.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
+      "--with-lapack=lapack"
+      (if use64BitIdx then "--enable-64" else "--disable-64")
+    ]
     ++ stdenv.lib.optionals stdenv.isDarwin [ "--enable-link-all-dependencies" ]
     ++ stdenv.lib.optionals enableReadline [ "--enable-readline" ]
     ++ stdenv.lib.optionals stdenv.isDarwin [ "--with-x=no" ]
     ++ stdenv.lib.optionals enableQt [ "--with-qt=5" ]
     ++ stdenv.lib.optionals enableJIT [ "--enable-jit" ]
```
```suggestion
    ++ lib.optionals stdenv.isDarwin [ "--enable-link-all-dependencies" ]
    ++ lib.optionals enableReadline [ "--enable-readline" ]
    ++ lib.optionals stdenv.isDarwin [ "--with-x=no" ]
    ++ lib.optionals enableQt [ "--with-qt=5" ]
    ++ lib.optionals enableJIT [ "--enable-jit" ]
```

KarlJoad replied:
Ping @doronbehar and @7c6f434c.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
+    # Keep a copy of the octave tests detailed results in the output
+    # derivation, because someone may care
+    postInstall = ''
+      cp test/fntests.log $out/share/octave/${pname}-${version}-fntests.log || true
```
Is the build log in hydra not enough?

KarlJoad replied:
Ping @doronbehar and @7c6f434c.

SuperSandro2000 commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -90,121 +92,147 @@ let
-    platforms = if overridePlatforms == null then
-      (with stdenv.lib; platforms.linux ++ platforms.darwin)
-    else overridePlatforms;
+    meta = {
+      homepage = "https://www.gnu.org/software/octave/";
+      license = stdenv.lib.licenses.gpl3Plus;
+      maintainers = with stdenv.lib.maintainers; [ raskin doronbehar ];
```
```suggestion
    meta = with lib; {
      homepage = "https://www.gnu.org/software/octave/";
      license = licenses.gpl3Plus;
      maintainers = with maintainers; [ raskin doronbehar ];
```

KarlJoad replied:
I can make this change, but because I am trying to work exclusively on octave packages before my school starts again. I am going to ping @7c6f434c and @doronbehar. I think having a separate PR/commit to handle that change would be better than tossing it into the mess of a PR this one is.

SuperSandro2000 commented on pkgs/development/interpreters/octave/hooks/octave-write-required-octave-packages-hook.sh:
```diff
@@ -0,0 +1,17 @@
+# Setup hook for writing octave packages that are run-time dependencies for
```
I think that filename could be shortened a bit.

KarlJoad replied:
Other than shortening the words themselves, the only other alternative is to remove the first `octave` and the `hook` at the end. I think the `hook` should stay, but the leading `octave` can be removed. Is that shorter enough, or should I attempt a complete re-name?

SuperSandro2000 replied:
To be honest this is just a nit but removing the first octave would be okay for me.

SuperSandro2000 commented on pkgs/development/interpreters/octave/wrap-octave.nix:
```diff
@@ -0,0 +1,18 @@
+, makeWrapper
+}:
+
+with lib;
```
Is this even used?

KarlJoad replied:
No, it wasn't. It was a left-over after reading through and copying a combination of Lua's and Python's `wrap-lang.nix` files. It has been removed.

SuperSandro2000 commented on pkgs/development/interpreters/octave/wrap.sh:
```diff
@@ -0,0 +1,131 @@
+
+    # Find all regular files in the output directory that are executable.
+    if [ -d "$dir" ]; then
+    find "$dir" -type f -perm -0100 -print0 | while read -d "" f; do
```
```suggestion
      find "$dir" -type f -perm -0100 -print0 | while read -d "" f; do
```

SuperSandro2000 commented on pkgs/development/interpreters/octave/wrap.sh:
```diff
@@ -0,0 +1,131 @@
+_addToOctavePath() {
+    local dir="$1"
+    # Stop if we've already visited this path.
+    if [ -n "${octavePathsSeen[$dir]}" ]; then return; fi
+        octavePathsSeen[$dir]=1
+        # addToSearchPath is defined in stdenv/generic/setup.sh. It has the effect
+        # of calling `export X=$dir/...:$X`.
+        addToSearchPath program_PATH $dir/bin
+
+        # Inspect the propagated inputs (if they exist) and recur on them.
+        local prop="$dir/nix-support/propagated-build-inputs"
+        if [ -e $prop ]; then
+        for new_path in $(cat $prop); do
+            _addToOctavePath $new_path
+        done
+    fi
```
```suggestion
    if [ -n "${octavePathsSeen[$dir]}" ]; then return; fi

    octavePathsSeen[$dir]=1
    # addToSearchPath is defined in stdenv/generic/setup.sh. It has the effect
    # of calling `export X=$dir/...:$X`.
    addToSearchPath program_PATH $dir/bin

    # Inspect the propagated inputs (if they exist) and recur on them.
    local prop="$dir/nix-support/propagated-build-inputs"
    if [ -e $prop ]; then
      for new_path in $(cat $prop); do
        _addToOctavePath $new_path
      done
    fi
```

KarlJoad commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -185,15 +196,25 @@ in mkDerivation rec {
 
-  passthru = {
+  passthru = rec {
+    octave = self;
```
Nope. I don't use it anywhere (at least that `nixpkgs-review` shows). So, I have removed it.

KarlJoad commented on pkgs/development/interpreters/octave/default.nix:
```diff
@@ -185,15 +196,25 @@ in mkDerivation rec {
+  passthru = rec {
+    octave = self;
     sitePath = "share/octave/${version}/site";
+    sitePackages = "share/octave/octave_packages";
```
Duplicate of `passthru.octPkgsPath`. This field has been removed in favor of the other (which is used in `build-env.nix`).

KarlJoad commented:
There are some packages that are quite old and haven't been updated for quite some time and are quite difficult to build. Is there a rule-of-thumb for how old a package should be before we consider dropping it?
For instance, the `secs2d` package hasn't received a release update since 2009. There are some others like this as well.

SuperSandro2000 commented:
> Is there a rule-of-thumb for how old a package should be before we consider dropping it?

If it is broken for a longer time, patching the build failure is non trivial and upstream has no commits for a year or two would be a pretty good indication to drop it.

> For instance, the secs2d package hasn't received a release update since 2009. There are some others like this as well.

If there is activity on the repo you can try to fetch the latest commit. If that is from 2009 or 2010 than I wouldn't feel bad removing it if it is not vital for the ecosystem.

KarlJoad commented:
> If there is activity on the repo you can try to fetch the latest commit. If that is from 2009 or 2010 than I wouldn't feel bad removing it if it is not vital for the ecosystem.

The sourceforge repo for that project doesn't even exist any more, so I think it's safe to assume `secs2d` (along with is sibling packages `secs1d` and `secs3d` are not actively maintained either. 

What would be the best way to remove them from the PR? A revert seems excessive. Rebase and drop the relevant commits?

doronbehar commented:
@KarlJoad In general, I think this PR should have only roughly these commits:

```
octave.pkgs: init

Add attributes to passthru of octave, add wrap-octave.nix etc.
```

And then:

```
octave.pkgs.symbolic: init at ...
octave.pkgs.io: init at ...
...
```

When you change something in the design of all of `octave.pkgs`, you can commit that with `--fixup` for the first `octave.pkgs: init` commit, [see this](https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash). The same goes for any other later change you do with a certain octave package.

All of the information about how this PR got evolved doesn't have much meaning IMO.

doronbehar commented on pkgs/development/interpreters/octave/build-octave-package.nix:
```diff
@@ -0,0 +1,103 @@
+
+    inherit meta;
+  } // { inherit patchPhase; };
+in lib.extendDerivation true passthru self
```
I think previously [@FRidh said](https://github.com/NixOS/nixpkgs/pull/108562#discussion_r556626751) that exactly this line - `extendDerivation` was a hack needed for python not needed here. I too think we should get by with just `stdenv.mkDerivation` as always.

KarlJoad replied:
Ahh... I don't quite understand how it is a hack, but OK. I just removed it. Testing and checking it now.

KarlJoad commented on pkgs/development/octave-modules/fits/default.nix:
```diff
@@ -0,0 +1,28 @@
+    license = licenses.gpl3Plus;
+    maintainers = with maintainers; [ KarlJoad ];
+    description = "Functions for reading, and writing FITS (Flexible Image Transport System) files using cfitsio";
+    # Marked this way until KarlJoad gets cfitsio as a runtime dependency.
```
In the initial commits, I just got all the packages into Nix. During that time, I didn't have a robust way of giving runtime-dependencies needed by packages to Octave. Now that I do, I have a branch (not in this PR, yet) where I fix up most of the packages that have runtime dependencies and remove the `broken = true;` flag. I am still working out some kinks in those before I pull them (and squash them) back into here.

There are still some packages that are still marked with `broken = true` because their runtime dependency doesn't exist in nixpkgs yet.

KarlJoad commented on pkgs/top-level/octave-packages.nix:
```diff
@@ -0,0 +1,261 @@
+      # I am still not super sure about this
+      dolfin = python.pkgs.fenics;
+      fenics = python.pkgs.fenics;
+      pkg-config = pkgs.pkg-config;
```
That's done on the fix-broken-packages branch that I mention [here](https://github.com/NixOS/nixpkgs/pull/108562#discussion_r561332184). 

However, `fenics` isn't a named thing. This was a "placeholder" for now, because I needed to remind myself what was going on here. `fenics` is a library, which is (sort of) in nixpkgs. Right now, `fenics` is only defined and built for the Python package `fenics`, which provides python bindings to the `fenics` **library**. However, I need access to the `fenics` **library**, but since that has no name in nixpkgs (the definition of the `fenics` library for python is in a let-binding), I can't access the library.

The definition of the fenics library can either be moved somewhere else, and be made a distinctly named entity (which I would prefer). Or, I can re-define fenics here again.

KarlJoad commented on pkgs/development/octave-modules/general/default.nix:
```diff
@@ -0,0 +1,28 @@
+
+  nativeBuildInputs = [
+    pkg-config
+    nettle
```
In that case, what should be in `nativeBuildInputs` rather than `buildInputs`? I'm not quite sure where certain tools are supposed to go. I know that the main difference comes down to cross-compilation, but am unsure beyond that. (This goes for all the other comments (below) regarding this particular issue.

KarlJoad commented:
I have also managed to fix almost all of the packages that were marked as broken before.
I will need to open PRs for the last 4 that are broken.

doronbehar commented:
Well done for all the work so far @KarlJoad. Overall this looks good to me though I will not be able to test this thoroughly until the end of the month. I still think though that the commit history could be squashed a bit further. The following could be squashed into one `octave.pkgs: init` IMHO:

```
octave: add predicate to check if a package is an octave package
octave: define self for this derivation
octave-packages: add octavePackages to all-packages.nix
octave-packages: use new requiredOctavePackages build list
octave-packages: write a withPackages function
octave.pkgs: unpack octave packages 
octave.pkgs: use inherit packages (implicitly when possible)
octave.pkgs: use nativeBuildInputs and buildInputs appropriately
octave.pkgs: with stdenv.lib; in meta -> with lib;
octave.pkgs: write buildOctavePackage function
octave.pkgs: write function build environment with octave packages
octave: remove second evaluation of pkgs inside octave's build
octave: start a wrapper.nix, mirroring python's
octave: start working on building environment to use everything with 
octave: start work on top-level file for octave packages
```

You can write some implementation details inside the commit message I guess, but besides that I personally see them all as part of the same thing.

Also this commit:

```
octave.pkgs: change src.url octave.sourceforge -> mirror://sourceforge
```

Could be part of the init's of every new octave package added.

KarlJoad commented:
How should I go about running tests, if at all? 
Octave requires a package be built **and** installed before being able to run tests on it.
This means that there is no way to run the tests during the compilation of the packages. I already have a script that runs tests for all packages that have been installed, but I'm not sure that we want to run tests on all packages specified every time someone creates an new environment with new octave packages.

Is there a way to have these tests always be run on hydra, and the end-user can specify if they want to run tests on their package/environment?

ravenjoad committed:
octave.pkgs: init

Heavily based on Python's packages set.

doronbehar commented:
> How should I go about running tests, if at all?

What I meant by "tests" is just running an octave interpreter with those packages available, and testing a few examples from these packages' docs - very manual, not necessarily putting it into the Nix expressions.

> I already have a script that runs tests for all packages that have been installed, but I'm not sure that we want to run tests on all packages specified every time someone creates an new environment with new octave packages.

I agree. Could you perhaps share that script?

> Is there a way to have these tests always be run on hydra, and the end-user can specify if they want to run tests on their package/environment?

Perhaps that would be nice to get done in the future, but it shouldn't block the PR IMO.

------

Besides that, everything looks good to me now :] with the commit history rewritten. I'll test some packages manually at the end of the month and I'll give my approval if all looks well. Thanks for all the hard work @KarlJoad !

KarlJoad commented:
With Octave 6.x a new command `pkg test <package name>`, or `pkg("test", "package-name")`. was created. It runs all the tests in a package's test suite. 
This test suite is included when the author of the package is writing it, so it tests most major functionalities (some packages have lackluster testing suites, but that's a different issue).

The "script" to test all packages currently installed into this Octave environment.
```octave
pkgs = pkg ("list");
for pkgNum = 1:numel (pkgs)
  thisPkg = pkgs{pkgNum}.name
  pkg ("load", thisPkg)
  pkg ("test", thisPkg)
  rename ("fntests.log", strcat (thisPkg, "-fntests.log"))
endfor
```

FRidh commented on pkgs/development/octave-modules/symbolic/default.nix:
```diff
@@ -0,0 +1,44 @@
+, lib
+, fetchurl
+, python
+, python2Packages
```
avoid passing in both the interpreter and the package set. Choose one. Both contain each other.

KarlJoad replied:
I actually have to do that. The python version that gets passed in is from Octave's default.nix, which is python3. But, symbolic is built assuming an older version of sympy, which is only contained in python2's packages. However, symbolic requires python3 be present.

FRidh reviewed (approved):
This looks good to me and I think we should merge it. Other (minor) changes can always be made after.

doronbehar commented:
Thanks once more @KarlJoad for all the work. I took the liberty to shorten the commit messages even further to include only necessary information. I tested `symbolic` which I intend to use in my default `octave` interpreter and I also tested `strings` (as a random example) and it proved to be functional enough. Will merge once CI is green! :tada: .

The next step is to document how Octave packages are handled in Nixpkgs... :book:.

Misterio77 commented:
Hey, @doronbehar. Sorry for the ping, and i know this isn't exactly the right place to ask, but i've looked around and couldn't get `symbolic` working using nix at all.

Should `nix-shell -p octave octavePackages.symbolic` be enough to get octave to load symbolic, or am i missing something?

I'd love to help document this if at all possible, too.

Misterio77 commented:
Got it working using `octave.withPackages`, just in case someone stumbles upon this

```nix
{ pkgs ? import <nixpkgs> {} }:
pkgs.mkShell {
  nativeBuildInputs = with pkgs; [
    (octave.withPackages (ps: with ps; [symbolic]))
  ];
}
```

KarlJoad commented:
@Misterio77, while technically possible open a shell with the octave package completely standalone, there is some trickery with text-based database files when using `withPackages` that is required to make packages usable with the actual octave package. If you specify _just_ the package, then the package will be downloaded, compiled, and saved as an output tarball for later installation. If you want to actually use the package, then you must specify `octave.withPackages`, much like you must with Python.

Misterio77 commented:
That makes a lot more sense! Thanks for helping a newbie understand it haha :))
